src/Service/CartLinesService.php

```diff
@@ -73,17 +74,20 @@ public function getCartLines($amount, $paymentFee, Cart $cart)
         $remaining = round($amount, $apiRoundingPrecision);
         $shipping = round($cart->getTotalShippingCost(null, true), $apiRoundingPrecision);
         $cartSummary = $cart->getSummaryDetails();
-        $cartItems = $cart->getProducts();
+        $cartItems = $cart->getProductsWithSeparatedGifts();
         $wrapping = Configuration::get('PS_GIFT_WRAPPING') ? round($cartSummary['total_wrapping'], $apiRoundingPrecision) : 0;
-        $remaining = round($remaining - $shipping - $wrapping, $apiRoundingPrecision);
-        $totalDiscounts = isset($cartSummary['total_discounts']) ? round($cartSummary['total_discounts'], $apiRoundingPrecision) : 0;
+        $totalDiscounts = $cart->getDiscountSubtotalWithoutGifts();
+        $remaining = round($remaining + $shipping + $wrapping - $totalDiscounts, $apiRoundingPrecision);
 
```

Comment:
I assume those numbers are floats? In PrestaShop its better to use Number class for any sort of math as floats sometimes produce inaccuracies.

https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems

Reply:
forgot about this one :D